rclcpp/include/rclcpp/node_interfaces/node_graph.hpp

```diff
@@ -117,9 +118,35 @@ class NodeGraph : public NodeGraphInterface
   size_t
   count_graph_users() override;
 
+  RCLCPP_PUBLIC
+  std::vector<rclcpp::TopicInfo>
+  get_publishers_info_by_topic(
+    const std::string & topic_name,
+    bool no_mangle = false) const override;
+
+  RCLCPP_PUBLIC
+  std::vector<rclcpp::TopicInfo>
+  get_subscriptions_info_by_topic(
+    const std::string & topic_name,
+    bool no_mangle = false) const override;
+
 private:
   RCLCPP_DISABLE_COPY(NodeGraph)
 
+  typedef rcl_ret_t (* rcl_get_info_by_topic_func_t)(
+    const rcl_node_t * node,
+    rcutils_allocator_t * allocator,
+    const char * topic_name,
+    bool no_mangle,
+    rmw_topic_info_array_t * info_array);
+
+  std::vector<rclcpp::TopicInfo>
+  get_info_by_topic(
```

Comment:
This can be moved as a free function in `node_graph.cpp` (it will need a `NodeBaseInterface` argument, instead of `this`).
`rcl_get_info_by_topic_func_t` can be replaced with a template parameter, as commented above.

Reply:
This fits nicely with my [comment](https://github.com/ros2/rclcpp/pull/960/files/9866d7e16ed76463b70432d1e30d67057183dc64#r366542431) as well: `type` can be replaced with template parameter as well.

Reply:
Is this still a thing? (conversation was not resolved, so I'm unsure)

Reply:
I think this should be resolved.